### ISSUE TITLE
There are no black bars (gutters) visible on the screen (HTML/GWT)

### ISSUE THREAD

**Issue by Cukierek** [libgdx/libgdx#4762]
#### Issue details
The game does not construct gutters (black bars) when resizing browser window (applied ExtendViewport with world from 720 480 to 864 552). gif: http://imgur.com/a/VkbtX
The proportions are fine as long as window size is within safezone. The gutters should appear when screen size is out of safezone. The game stretches the view instead (same code works fine on desktop and android).

#### Reproduction steps/code

Any game screen with this HtmlLauncher ...

````java
public class HtmlLauncher extends GwtApplication {
    static final int WIDTH = 864;
    static final int HEIGHT = 552;
    static HtmlLauncher instance;

    @Override
    public GwtApplicationConfiguration getConfig() {
        GwtApplicationConfiguration config = new GwtApplicationConfiguration(Window.getClientWidth(), Window.getClientHeight());

        Element element = Document.get().getElementById("embed-html");
        VerticalPanel panel = new VerticalPanel();
        panel.setWidth("100%");
        panel.setHeight("100%");
        panel.setHorizontalAlignment(HasHorizontalAlignment.ALIGN_CENTER);
        panel.setVerticalAlignment(HasVerticalAlignment.ALIGN_MIDDLE);
        element.appendChild(panel.getElement());
        config.rootPanel = panel;

        return config;
    }

    @Override
    public ApplicationListener getApplicationListener() {
        instance = this;
        setLogLevel(LOG_NONE);
        setLoadingListener(new LoadingListener() {
            @Override
            public void beforeSetup() {

            }

            @Override
            public void afterSetup() {
                scaleCanvas();
                setupResizeHook();
            }
        });
        return new HiddenObjectGame();
    }

    void scaleCanvas() {
        Element element = Document.get().getElementById("embed-html");
				
        int innerWidth = getWindowInnerWidth();
        int innerHeight = getWindowInnerHeight();

        NodeList<Element> nl = element.getElementsByTagName("canvas");
		
        if (nl != null && nl.getLength() > 0) {
            Element canvas = nl.getItem(0);
            canvas.setAttribute("width", "" + innerWidth + "px");
            canvas.setAttribute("height", "" + innerHeight + "px");
            canvas.getStyle().setWidth(innerWidth, Style.Unit.PX);
            canvas.getStyle().setHeight(innerHeight, Style.Unit.PX);
            canvas.getStyle().setTop(0, Style.Unit.PX);
            canvas.getStyle().setLeft(0, Style.Unit.PX);
            canvas.getStyle().setPosition(Style.Position.ABSOLUTE);
        }
    }

    native int getWindowInnerWidth() /*-{
        return $wnd.innerWidth;
    }-*/;

    native int getWindowInnerHeight() /*-{
        return $wnd.innerHeight;
    }-*/;

    native void setupResizeHook() /*-{
        var htmlLauncher_onWindowResize = $entry(@com.imaginalis.hog.client.HtmlLauncher::handleResize());
        $wnd.addEventListener('resize', htmlLauncher_onWindowResize, false);
    }-*/;

    public static void handleResize() {
        instance.scaleCanvas();
    }

	@Override
	public ApplicationListener createApplicationListener() {
		return getApplicationListener();
	}
}
````

and with this viewport in core:
````java
viewport = new ExtendViewport(720, 480, 864, 552);
gameStage = new Stage(viewport);
````
And again, this solution works fine on other platforms. I realize I could do something wrong. If so, point it out bluntly.

EDIT

I'm putting Game and Screen class for your convenience. Just copy and paste these two into the core project:

Game
````java
public class ViewportTestGame extends Game {
	
	public SpriteBatch spriteBatch;
	
    @Override
    public void create() {
        spriteBatch = new SpriteBatch();
		this.setScreen(new ViewportTestScreen(this));
    }
    
    @Override
    public void render() {
        super.render();
    }
    
    @Override
    public void dispose() {
        spriteBatch.dispose();
    }
}
````

Screen 

````java
public class ViewportTestScreen implements Screen {
	
	private Game game;
	private Stage gameStage;
	private Image backgroundTestImage;
	private SpriteBatch spriteBatch;

	public ViewportTestScreen(final ViewportTestGame game) {
		this.game = game;
		this.backgroundTestImage = new Image(new Texture(Gdx.files.internal("background_test3.png")));
		this.gameStage = new Stage(new ExtendViewport(720, 480, 864, 552));
		this.gameStage.addActor(backgroundTestImage);
		spriteBatch = new SpriteBatch();
		Gdx.input.setInputProcessor(gameStage);
		// if you need virtual 0,0 point...
//		float virtualZeroX = (864 - gameStage.getViewport().getWorldWidth()) / 2f;
//		float virtualZeroY = (552 - gameStage.getViewport().getWorldHeight()) / 2f;
	}

	@Override
	public void show() {
		gameStage.getViewport().getCamera().position.set(864 / 2f, 552 / 2f, 0);
	}

	@Override
	public void render(float delta) {
		Gdx.gl.glClearColor(0, 0, 0, 1);
		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
		gameStage.act();
		gameStage.draw();
	}

	@Override
	public void resize(int width, int height) {
		gameStage.getViewport().update(width, height);
		gameStage.getViewport().getCamera().position.set(864 / 2f, 552 / 2f, 0);
	}

	@Override
	public void pause() {
	}

	@Override
	public void resume() {
	}

	@Override
	public void hide() {
	}

	@Override
	public void dispose() {
	}
}
````
In desktop project:

````java
public static void main (String[] arg) {
		LwjglApplicationConfiguration config = new LwjglApplicationConfiguration();
				int width = 1366;
				int height = 768;
                config.width = width;
                config.height = height;
                config.resizable = true;
				new LwjglApplication(new ViewportTestGame(), config);
````

It's dangerous to go alone. Take this:

![background_test3](https://cloud.githubusercontent.com/assets/26262167/26452856/d293a0d6-4161-11e7-9178-5f75206029e9.png)

... and put it into the assets folder

#### Version of LibGDX and/or relevant dependencies
1.9.6, no additional dependencies

#### Please select the affected platforms
- [ ] Android
- [ ] iOS (robovm)
- [ ] iOS (MOE)
- [x] HTML/GWT
- [ ] Windows
- [ ] Linux
- [ ] MacOS

**Comment by Tom-Ski** [comment]
This seems to be a problem with your resizing code. Have you checked that resize is being called? If not, then debug the size of the canvas.

**Comment by MrStahlfelge** [comment]
Yes, here is a working resizing code.

```
    private GwtApplicationConfiguration cfg;

    @Override
    public GwtApplicationConfiguration getConfig() {
        // reduce by 20 to avoid scrolling in iframes
        int w = Window.getClientWidth() - 20;
        int h = Window.getClientHeight() - 20;
        cfg = new GwtApplicationConfiguration(w, h);
        Window.enableScrolling(false);
        Window.setMargin("0");
        Window.addResizeHandler(new ResizeListener());
        cfg.preferFlash = false;
        return cfg;
    }

    class ResizeListener implements ResizeHandler {
        @Override
        public void onResize(ResizeEvent event) {
            int width = event.getWidth();
            int height = event.getHeight();
            getApplicationListener().resize(width, height);
            Gdx.graphics.setWindowedMode(width, height);
            Gdx.gl.glViewport(0, 0, width, height);

            Window.scrollTo((cfg.width - width) / 2, (cfg.height - height) / 2);

        }
    }
```

@Tom-Ski if you assure me to pull it into the master and tell me where the template is, I would be glad to contribute this into the project.

**Comment by Tom-Ski** [comment]
This would make all games take up the whole window size (- padding), no?

**Comment by MrStahlfelge** [comment]
Exactly, same behaviour than Android and Desktop. 

Normally it is not the browser window, because HTML games are living in an iframe by the hosing site in most cases. So the game would fill the whole iframe, yes.

**Comment by Tom-Ski** [comment]
Sounds good.  Template file is here, feel free to contribute https://github.com/libgdx/libgdx/blob/master/extensions/gdx-setup/src/com/badlogic/gdx/setup/resources/html/src/HtmlLauncher

**Comment by MrStahlfelge** [comment]
Just to drop a note on the current state. Despite I used the code above for all my games, I discovered when I wanted to contribute that it does not work correctly when the game is resized to a smaller size than the initial size: Clicks are not registered at the correct position anymore. I never noticed that because normally all users resize the game to a size bigger than initial.

I have another code here that works like charm on this, but it has the disadvantage that the preloader logo is not centered on start. I am working on this.
